### DISCUSSION TITLE
Automatic Created and Updated columns

### DISCUSSION THREAD

**Discussion by peachp** [supabase/supabase#6741] [Feature Requests]
# Feature request

## Is your feature request related to a problem? Please describe.
No specific problem right now, but it is a very common and useful feature.

## Describe the solution you'd like
When creating a table allow to add "Created" and "Updated" columns, which will be populated by the system accordingly

## Describe alternatives you've considered
Do in client side code for each table for each INSERT / UPDATE

**Comment by awalias** [comment]
Could be a good one @joshenlim

**Comment by kiwicopple** [comment]
I think for this one we would need to add triggers - https://x-team.com/blog/automatic-timestamps-with-postgresql/

We could "auto populate" the field on the dashboard, but it would be missing values when a row is programmatically inserted (via the API). Will the trigger approach work for you @peachp ?

**Comment by peachp** [comment]
@kiwicopple I think populating it on the lowest level possible, would be great. So it won't matter whether I insert / update a row via your web GUI, API, SQL or maybe even when connected to the DB directly. Though at least via API and web GUI would be good.

**Comment by kiwicopple** [comment]
Hey @peachp , we now have the `moddatetime` extension, so you can run this

```sql

create extension if not exists moddatetime schema extensions;

-- assuming the table name is "todos"
create trigger handle_updated_at before update on todos 
  for each row execute procedure moddatetime (updated_at);
```

I'll add this to the quickstart snippets

**Reply by fvaldes33** [reply]
What would this look like if we wanted to update the "todos" table when "todo_items" were updated? Is there a way to leverage the joins in the trigger?

**Comment by peachp** [comment]
@kiwicopple
I could do this for tables where I really need it, though it would be great to have a GUI where the two columns (e.g. `updated_at` and `created_at`) can be activated when creating a new table or editing it under Database.
Or maybe even allow enabling it globally for all tables!

> We could "auto populate" the field on the dashboard, but it would be missing values when a row is programmatically inserted

No no, of course it wouldn't make sense to auto-populate rows only when inserting / updating via dashboard. To clarify, when I wrote
> When creating a table allow to add "Created" and "Updated" columns

By that I meant when creating a new table under Database, like here under "Include primary key" for example:
![image](https://user-images.githubusercontent.com/47209189/103882300-808d5680-50d3-11eb-8cf3-b081c9f2da24.png)

And when you edit the table, the GUI should also show whether these two fields are being auto-populated, because it might be risky to judge only by the data itself, whether they are populated: if you run `create trigger... updated_at` then do `drop trigger... updated_at` for some reason, you would have to remember (or run some SQL to see active triggers), because you would still have `update_at` values in your data, but they will be outdated.

Anyways!
What you provided is already a good start. I guess you could make it more user-friendly at some point.

**Comment by kiwicopple** [comment]
Yeah we need much better "trigger" support in the dashboard. It's coming !

> the GUI should also show whether these two fields are being auto-populated

This one will be a bit harder if they are populated by triggers, but let us brainstorm this one. ~~I'll convert this one to a Discussion for now since it's not a breaking issue. It will be better as the community can then search for it in the future~~

It seems I can't convert issues to discussions any more 🤔

**Comment by peachp** [comment]
Another way would be to have a special type when you add a field. Have a look how Airtable does it.

They solved it pretty good I think: you can add or remove such fields at any time, even after you added records. Implies that they always maintain something like `created_at` and `updated_at` under the hood, so when you add fields of this type, you are just "surfacing" them, so that your pre-existing records will have it too.

For `updated_at` they go even further: you can chose to update it only when certain fields change. This allows logging / auditing, but also to build some neat business logic on top of it (which we actually do).

I imagine this could be low priority on your list because of relatively high complexity / low value (?), so no rush. But it would be great to have it at some point. This is something I've seen in a few mature database services / ORMs / frameworks.

![image](https://user-images.githubusercontent.com/47209189/104025520-acd0d200-51bc-11eb-98ef-2e8d4ab73590.png)

![image](https://user-images.githubusercontent.com/47209189/104025786-0c2ee200-51bd-11eb-8495-b6fe7682a51c.png)

**Reply by nkeat12** [reply]
This would be an awesome add

**Comment by kiwicopple** [comment]
For anyone stumbling upon this, we have a small workaround. There is a script in the SQL editor called "Automatically update timestamps" which will automatically update a timestamp

<img width="470" alt="image" src="https://user-images.githubusercontent.com/10214025/114698248-78d3f280-9d51-11eb-8ee2-8a2419901931.png">

**Reply by s-kennedy** [reply]
I'm trying to use this script, and I get the following error:
`Failed to run sql query: permission denied to create extension "moddatetime"`

I also tried manually enabling the extension but I get the same permissions error. I have "developer" access on the project. Is this something the project owner needs to do?

**Comment by Polyterative** [comment]
Works well for me

**Comment by surjithctly** [comment]
Just wondering how do this work for Created_at, currently it does not add column when creating new rows. 

I wanted to add a column created_at when new row is created. 

Any help?

**Comment by kiwicopple** [comment]
for that @surjithctly , you can just create a column `created_at timestamptz default now()`

![image](https://user-images.githubusercontent.com/10214025/124708705-37587b00-df2d-11eb-8dd9-5e591ba56308.png)

**Reply by jarvismartin** [reply]
I have these settings, but when I insert via the JS client, I get an error for not providing a value for created_at.
`{
    "code": "23502",
    "details": null,
    "hint": null,
    "message": "null value in column \"created_at\" of relation \"reviews\" violates not-null constraint"
}`

What is the right way to do this? I just want a timestamp each time I create a new record.

**Comment by kiwicopple** [comment]
We now auto-add the `created_at` column

<img width="670" alt="image" src="https://user-images.githubusercontent.com/10214025/147556464-0faade6c-232c-4ab8-a00e-6fe5b3b4065e.png">


I'll convert this one to a discussion, as i think it will be useful in the future for anyone else who wants to learn about the `updated_at` triggers

**Comment by riderx** [comment]
For everyone who need `updated_at`
Full procedure :
- go to extensions and allow `moddatetime`
- add a new column in the table you need it with, type `timestamptz`
- got to SQL editor and create a new one called `Set updated_at autoupdate` copy in it 
```sql
create trigger handle_updated_at before update on [TABLENAME] 
  for each row execute procedure moddatetime (updated_at);
 ```
 where you replace TABLENAME by the name of your table
 - run the command 
 - go to your table and do a modif after few seconds you should see the new value appear.
 
 Enjoy ❤️

**Reply by t6adev** [reply]
> go to extensions and allow moddatetime

We could execute a SQL to turn it on, instead.
```
create extension moddatetime;
```

https://github.com/supabase/supabase/discussions/6741#discussioncomment-2713242

**Comment by MildTomato** [comment]
Should we be looking to add this in as something that can be turned on via some simple UI ?
Rather than running SQL? 

cc @kiwicopple @saltcod

**Reply by MentalGear** [reply]
Yup, automatic update time would be nice.

**Comment by surjithctly** [comment]
@MildTomato  Yes Please

**Comment by kiwicopple** [comment]
yup, good one

**Comment by Faridalim** [comment]
Any update on this feature? Would be awesome if we can turn on a feature for this using GUI

**Reply by bennik88** [reply]
Would be a welcomed feature instead of always using triggers.

**Reply by saltcod** [reply]
Hey all. I just added this to an internal board. Hope to get it prioritized soon. Thanks!

**Reply by ZetiMente** [reply]
Thanks @saltcod

**Reply by riderx** [reply]
thanks ! i just found out today i forgot to add manually the trigger on some new table so all history is loss :/

been able to see if trigger are present also can be great

**Reply by bennik88** [reply]
What is the best practice for securing the `created_at` value from being updated by the authenticated user? As the user usually has RLS to update the table which also has the `created_at`column. When using triggers the `updated_at` will overwrite values the user tried to manually update so it protects the value, but with `created_at` this won't work.

**Reply by point-source** [reply]
> What is the best practice for securing the `created_at` value from being updated by the authenticated user? As the user usually has RLS to update the table which also has the `created_at`column. When using triggers the `updated_at` will overwrite values the user tried to manually update so it protects the value, but with `created_at` this won't work.

You could actually overwrite created_at as well in the same trigger as updated_at. Since for update operations you can access the old and new record values, just take the created_at value from the old record and write it over the new one.

**Reply by bennik88** [reply]
That's an interesting approach, thanks. But the user could still mess up the id of the column let's say it's a cars table. He would still be able to change the car `id`. So should I do that for every column I want to prevent changes? Sounds not too great as I can't easily see what has been written in the trigger and therefore do not check security easily.

**Reply by point-source** [reply]
Sure. You could also raise an exception if you want to inform the user that this is not allowed. See here: https://stackoverflow.com/questions/56082230/postgres-trigger-to-avoid-update-in-a-column

**Reply by riderx** [reply]
You can also Check in RLS add `created_at IS NULL` and then refuse change who include data in the field
Edit this is not possible to do that with RLS, but you can do VIEW https://supabase.com/blog/postgresql-views

**Reply by bennik88** [reply]
@riderx I think it's not possible to check with RLS. And views don't help in a case where a user can have multiple contacts and is able to sort his data by `created_at` as I want to enable the user to select the data but restrict updating the column `created_at`.

**Reply by bennik88** [reply]
It seems like the below function in combination with a trigger on INSERT won't work to protect the user from overwriting `created_at` or `id` on INSERT. This pattern seems to only work for UPDATE. It seems that `NOW()` isn't populated before the TRIGGER runs. This is weird as the user can overwrite `NOW()` but in TRIGGER, BEFORE INSERT it seems to be not yet available. 
```sql
CREATE OR REPLACE FUNCTION validate_insert()
RETURNS TRIGGER AS
$$
BEGIN
  -- always reset id created_at and owner_id to the value already stored
  NEW.id := OLD.id;
  NEW.created_at := OLD.created_at;
  NEW.owner_id := OLD.owner_id;
  RETURN NEW;
END;
$$
LANGUAGE plpgsql;
```

I've also tried it using this, which also doesn't work. The user can still overwrite the three fields:
```sql
BEGIN
  IF NEW.id <> NULL
  OR NEW.created_at <> NULL
  OR NEW.last_updated_at <> NULL
   THEN
    RAISE EXCEPTION 'Cannot set id, created_at or last_updated_at';
    END IF;
    RETURN NEW;
END;
```

**Reply by riderx** [reply]
Yes sorry for that that not possible I did tried also. the only way is to use a view https://www.guru99.com/postgresql-view.html

**Comment by jdgamble555** [comment]
Please see my article on covering ALL security use cases, and can be reusable. Moddatetime can still be overwritten on its own.

https://code.build/p/XUPzns9ykZj4Abcazg1kt5/supabase-date-protection-trigger-functions

J

**Comment by ChuckJonas** [comment]
Not to derail the conversation, but it would be amazing to have something similar for `created_by` and `last_modified_by` which references a user.

**Comment by 100kristine** [comment]
Is this still not in the GUI? Didn't see it at a quick docs/google search either.

**Comment by thubv** [comment]
<img width="1120" alt="image" src="https://github.com/supabase/supabase/assets/7010816/950802d7-1e65-4a6b-964e-50c15c8546d1">
Please consider using updated_at as a default method.

**Comment by bodadotsh** [comment]
Just to check, this is still not available in UI right? 

<img width="642" height="182" alt="image" src="https://github.com/user-attachments/assets/672bb494-058f-4d5f-9d71-2ea7aa1eb850" />

The best way is still to execute the "Automatically update timestamps" script in "SQL Editor > Templates"? 

<img width="1048" height="480" alt="image" src="https://github.com/user-attachments/assets/687e837e-ca41-4a14-964d-7a7c64e9d6ad" />
